### DISCUSSION TITLE
How to get rid of question marks in openapi.json paths when using `/?` to fix the trailing slash issue?

### DISCUSSION THREAD

**Discussion by stianlagstad** [fastapi/fastapi#7560] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

How can I get rid of question marks in openapi.json paths when using `/?` to fix the trailing slash issue?

We're seeing [this exact thing](https://github.com/tiangolo/fastapi/issues/51#issuecomment-489381899), that is the question marks that we've added to "fix" the trailing slash issue appears in the generated openapi.json paths.

### Additional context

Image from the other issue:

![docs with question mark](https://user-images.githubusercontent.com/1664041/57187353-0a2a6100-6ebb-11e9-85b8-0aa97d405aab.png)

This happens when you have endpoints like these:
```
@router.get("/budgets/?")
@router.post("/budgets/?")
```

**Comment by jorgerpo** [comment]
I don't get it.
You are putting question marks in your endpoints but don't want them?

**Comment by stianlagstad** [comment]
The question marks makes sure that both the path `/budgets` and the path `/budgets/` are possible to call with a POST request. Please see [the other issue that I linked to](https://github.com/tiangolo/fastapi/issues/51#issuecomment-489381899).

**Comment by chris-allnutt** [comment]
You can extend the api and overwrite each endpoint as part of that operation.  I would be concerned that this is a band-aid over a known issue but will get you what you want

https://fastapi.tiangolo.com/advanced/extending-openapi/

You should be able to iterate thru what get's initially generated and alter the lines you want removing the ?

**Comment by chris-allnutt** [comment]
You could potentially put something like nginx in front and do conditional re-writes which is how I would solve this problem personally removing the '?', so in the future if there's a change or fix released it won't be an issue.

**Comment by luckydonald** [comment]
Wouldn't you just do 
```py
@router.get("/budgets")
@router.get("/budgets/")
```
?

That's what I did with flask routes.

**Comment by stianlagstad** [comment]
Thank you for the responses! And sorry for being late to respond myself.

A colleague of mine created this solution:
```
from fastapi import FastAPI
from fastapi.openapi.utils import get_openapi


def get_openapi_fix_question_marks(app: FastAPI):
    def _get_openapi():
        if app.openapi_schema:
            return app.openapi_schema
        openapi_schema = get_openapi(
            title=app.title,
            version=app.version,
            description=app.description,
            routes=app.routes,
        )
        openapi_schema["paths"] = {
            k: v
            for k, v in map(
                lambda key: (key.rstrip("?"), openapi_schema.get("paths", {}).get(key)),
                openapi_schema.get("paths", {}).keys(),
            )
        }
        app.openapi_schema = openapi_schema
        return app.openapi_schema

    return _get_openapi
```

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

Just for completeness, you should be able to declare your path operations with just `/something/` and have the version not ending in `/` redirected automatically, that would probably be the default for most cases.
